### PR TITLE
allow to definie static fees for transactions. Fixes #137

### LINKED ISSUE #137: absolute fee
I need to be able to construct transactions with an absolute value for the fee. To come close, I could construct a tx, measure the size, and use some reverse calculation to approximate the FeeRate. But that might not be exact enough.
I am willing to implement this feature myself, and make a pull request. Which of the following directions should I go:
`wallet.create_tx(TxBuilder::with_recipients(vec![(address.script_pubkey(), satoshis)]).fee_rate(FeeRate::absolute(fee))`
or
`wallet.create_tx(TxBuilder::with_recipients(vec![(address.script_pubkey(), satoshis)]).fee_absolute(fee)`
I would opt for the latter.

### PR DESCRIPTION
**Pull request by ulrichard**

### CODE REVIEW

File: src/wallet/tx_builder.rs [hunk 1]
```diff
@@ -303,6 +315,14 @@ impl<D: Database, Cs: CoinSelectionAlgorithm<D>> TxBuilder<D, Cs> {
             phantom: PhantomData,
         }
     }
+
+    /// Returns true if an absolute fee was specified
+    pub fn has_absolute_fee(&self) -> bool {
+        if self.fee_policy.is_none() {
+            return false;
+        };
+        matches!(self.fee_policy.as_ref().unwrap(), FeePolicy::FeeAmount(_))
```
Comment (afilini) [inline src/wallet/tx_builder.rs]: If you need to keep this code, it can probably be refactored to just:

```suggestion
        matches!(self.fee_policy, Some(FeePolicy::FeeAmount(_)))
```

File: src/wallet/mod.rs [hunk 2]
```diff
@@ -299,7 +299,7 @@ where
             output: vec![],
         };
 
-        let fee_rate = builder.fee_rate.unwrap_or_default();
+        let fee_rate = get_fee_rate(&builder.fee_policy);
```
Comment (afilini) [inline src/wallet/mod.rs]: I think the `fee_rate` should be set to `0.0` when there's an absolute fee (the default is `1.0`), otherwise in some edge cases the coin selection could fail because it would be considering the extra cost for each input added while in fact we don't care about it since we override the `fee_amount` later.

Also, the `fee_amount` should be overridden before the coin selection is done, so that the coin selection can properly take into account that amount of fees.

File: src/wallet/mod.rs [hunk 3]
```diff
@@ -393,6 +393,14 @@ where
         };
 
         let mut fee_amount = fee_amount.ceil() as u64;
+
+        if builder.has_absolute_fee() {
+            fee_amount = match builder.fee_policy.as_ref().unwrap() {
+                FeePolicy::FeeAmount(amount) => *amount,
+                _ => fee_amount,
+            }
+        };
```
Comment (afilini) [inline src/wallet/mod.rs]: See comment above, I think this should be moved up before the coin selection.

I would actually do the `fee_rate`/`fee_amount` initialization in one go. Something like:

```rust
let (fee_rate, mut fee_amount) = match builder.fee_policy.as_ref().unwrap_or(FeePolicy::FeeRate(FeeRate::default())) {
    FeePolicy::FeeAmount(amount) => (FeeRate::from_sat_per_vb(0.0), amount as f32),
    FeePolicy::FeeRate(rate) => (rate, 0.0),
};
```

In my opinion, it also makes sense to implement `Default` for `FeePolicy`, which would return `FeePolicy::FeeRate(FeeRate::default())`. With that you could then turn the `.unwrap(...)` above to just `.unwrap_or_default()`.

As a bonus, this would also remove the need for the `get_fee_rate()` function.

File: src/wallet/mod.rs [hunk 4]
```diff
@@ -485,9 +493,9 @@ where
         // the new tx must "pay for its bandwidth"
         let vbytes = tx.get_weight() as f32 / 4.0;
         let required_feerate = FeeRate::from_sat_per_vb(details.fees as f32 / vbytes + 1.0);
-        let new_feerate = builder.fee_rate.unwrap_or_default();
+        let new_feerate = get_fee_rate(&builder.fee_policy);
```
Comment (afilini) [inline src/wallet/mod.rs]: See comment on `create_tx()`, I would remove `get_fee_rate()` and do something similar here.

File: src/wallet/mod.rs [hunk 5]
```diff
@@ -485,9 +493,9 @@ where
...
 
-        if new_feerate < required_feerate {
+        if new_feerate < required_feerate && !builder.has_absolute_fee() {
```
Comment (afilini) [inline src/wallet/mod.rs]: This doesn't check whether the new absolute fee is actually high enough to "pay for its bandwidth".

File: src/wallet/mod.rs [hunk 6]
```diff
@@ -652,21 +669,36 @@ where
         details.sent = selected_amount;
 
         let mut fee_amount = fee_amount.ceil() as u64;
+        if builder.has_absolute_fee() {
+            fee_amount = match builder.fee_policy.as_ref().unwrap() {
+                FeePolicy::FeeAmount(amount) => *amount,
+                _ => fee_amount,
+            }
+        };
```
Comment (afilini) [inline src/wallet/mod.rs]: You probably won't need to re-override the amount if you set the `fee_rate` correctly to `0.0`.

File: src/wallet/mod.rs [hunk 7]
```diff
@@ -652,21 +669,36 @@ where
...
             tx.output.push(removed_updatable_output);
         } else if builder.send_all && !change_val_after_add.is_dust() {
-            removed_updatable_output.value = change_val_after_add;
-            fee_amount += removed_output_fee_cost;
+            if builder.has_absolute_fee() {
+                removed_updatable_output.value = change_val_after_add + removed_output_fee_cost;
+            } else {
+                removed_updatable_output.value = change_val_after_add;
+                fee_amount += removed_output_fee_cost;
+            }
```
Comment (afilini) [inline src/wallet/mod.rs]: Same as above, if the `fee_rate` is correctly set to `0.0` you shouldn't need to handle the absolute fees as a special case, and you should be able to just leave the code as it was before.

File: src/wallet/mod.rs [hunk 8]
```diff
@@ -1210,6 +1242,17 @@ where
     }
 }
 
+/// get the fee rate if specified or a default
+fn get_fee_rate(fee_policy: &Option<FeePolicy>) -> FeeRate {
```
Comment (afilini) [inline src/wallet/mod.rs]: This fn can be removed, see comments above

File: src/wallet/tx_builder.rs [hunk 9]
```diff
@@ -303,6 +315,14 @@ impl<D: Database, Cs: CoinSelectionAlgorithm<D>> TxBuilder<D, Cs> {
             phantom: PhantomData,
         }
     }
+
+    /// Returns true if an absolute fee was specified
+    pub fn has_absolute_fee(&self) -> bool {
```
Comment (afilini) [inline src/wallet/tx_builder.rs]: This method can probably also be removed if you implement the changes I proposed in `src/mod/wallet.rs`.

**Review by afilini — CHANGES_REQUESTED** [review]
I like the idea of having a `FeePolicy` enum, but I think there are a few things you should change.

As a minor nit, when you implement those changes you could also just squash everything into a single commit which would make the history a bit more clean.

Also, I haven't reviewed the tests, but I would try to add a couple more to test the edge cases in which the current code would probably break: things like asking for an absolute fee of 0 satoshi, asking for a very high absolute fee, trying to bump a tx with an absolute fee that is not high enough, etc.

File: src/wallet/mod.rs [hunk 10]
```diff
@@ -485,13 +493,28 @@ where
...
+            }
+            FeePolicy::FeeRate(rate) => {
+                if *rate < required_feerate {
+                    return Err(Error::FeeRateTooLow {
+                        required: required_feerate,
+                    });
+                }
+                *rate
+            }
+        };
```
Comment (afilini) [inline src/wallet/mod.rs]: I think this `match` can be merged with the one below that initializes `initial_fee`. I would do it mostly for consistency with `create_tx` and to avoid duplicating the `.unwrap_or(...)` logic

File: src/wallet/mod.rs [hunk 11]
```diff
@@ -1660,6 +1691,67 @@ mod test {
...
+                    .send_all(),
+            )
+            .unwrap();
+
+        assert_eq!(details.fees, 0);
+        assert_eq!(psbt.global.unsigned_tx.output.len(), 1);
+        assert_eq!(
+            psbt.global.unsigned_tx.output[0].value,
+            50_000 - details.fees
+        );
```
Comment (afilini) [inline src/wallet/mod.rs]: You can probably remove those lines since we expect it to panic above them

**Review by afilini — APPROVED** [review]
